Data Grid - Phase 2

tjscience opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->


## Description
Phase 2 of the Data Grid brings missing and new features as well as performance improvements.
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310 -->

## How Has This Been Tested?
Unit tests and testing through development.
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Change the filter operators to constants. ([Commit 2e486fa6](https://github.com/tjscience/MudBlazor/commit/2e486fa61b94a2e0ada81bf5c1f33c90bc474b38)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-73842852))
- [x] Handle nullable types in filters. ([Commit 4ed72d89](https://github.com/tjscience/MudBlazor/commit/4ed72d890b957c33641f6fb197cbf724d8a25838)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-74683233))
- [x] More built in filter data types. ([Commit 4ed72d89](https://github.com/tjscience/MudBlazor/commit/4ed72d890b957c33641f6fb197cbf724d8a25838)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-74683233))
- [x] Column chooser. ([Commit b36519e9](https://github.com/tjscience/MudBlazor/commit/b36519e97adaa14dfe535a777d996e948b09035a)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-73842635))
- [x] Add grouping. ([Commit b0ff18a2](https://github.com/tjscience/MudBlazor/commit/b0ff18a2f340617fe09ffa46e41f1a5d86639a15)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-75678285))
- [x] Built-in aggregation for footers. ([Commit 02721ef2](https://github.com/tjscience/MudBlazor/commit/02721ef22b92dc871cd06ca0250df3951923fbf0)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-73842600))
- [x] Clean up child components (i.e. HeaderCell, Cell and FooterCell) by removing unnecessary Parameters. Also, replaced the Cell component with a normal class to improve performance. ([Commit 574b2346](https://github.com/tjscience/MudBlazor/commit/574b2346aea0721f4d2f6255d70eea2795dc95fb)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-77329892))
- [x] DataGridEditMode.Cell mode to allow excel like editing by cell. ([Commit b36519e9](https://github.com/tjscience/MudBlazor/commit/b36519e97adaa14dfe535a777d996e948b09035a)) ([Project Card](https://github.com/tjscience/MudBlazor/projects/1#card-73842583))

<a href="https://www.loom.com/share/bd7500adae354b818ddc42266b6a7996">
    <p>MudDataGrid - More Built-In Filter Data Types - Watch Video</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/bd7500adae354b818ddc42266b6a7996-with-play.gif">
  </a>

<a href="https://www.loom.com/share/4317da39ba444e39a12f78ee1408c9eb">
    <p>MudDataGrid - Column Chooser Hide/Show Columns - Watch Video</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/4317da39ba444e39a12f78ee1408c9eb-with-play.gif">
  </a>

<a href="https://www.loom.com/share/39fc3d00ebc449e8b35591230033a7c3">
    <p>MudDataGrid - Data Aggregation - Watch Video</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/39fc3d00ebc449e8b35591230033a7c3-with-play.gif">
  </a>

<a href="https://www.loom.com/share/a0b6870e2f7f41e6a72f4028d18d1051">
    <p>MudDataGrid - Editing - Watch Video</p>
    <img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/a0b6870e2f7f41e6a72f4028d18d1051-with-play.gif">
  </a>

tjscience commented:
@Garderoben, please take a look when you have a moment. I still have some tests to write and I also want to better document how to do things. For example, using the CellContext class to fine tune behavior at a granular level.

Garderoben commented:
> @Garderoben, please take a look when you have a moment. I still have some tests to write and I also want to better document how to do things. For example, using the CellContext class to fine tune behavior at a granular level.

any specific thing you want me to take a look at or overal?

mckaragoz commented:
If you will add multiple sorting in this phase, we will be appreciated :)

neozhu commented:
> If you will add multiple sorting in this phase, we will be appreciated :)

specify different filter property  in each column, This feature is needed more.
like this
<img width="522" alt="image" src="https://user-images.githubusercontent.com/1549611/155881903-d4d593ca-d66f-42b5-bf83-0c2127cc7d26.png">

IngoManthey commented:
Provide a way to localize your texts.

tjscience commented:
@Garderoben Due to the breadth of functionality available in the data grid, I think that docs-wise it would be better to keep the small number of showcase examples that exist currently and then have a list of links to https://try.mudblazor.com/ to exemplify all of the features and real-world use cases. Having a single page for documentation on a component like this does not really suffice and will result in a massive page eventually. Take a look at the documentation for DataGrid in MUI for a reference. They have multiple pages for DataGrid docs, each dealing with a particular concept. https://mui.com/components/data-grid/

Garderoben commented:
> @Garderoben Due to the breadth of functionality available in the data grid, I think that docs-wise it would be better to keep the small number of showcase examples that exist currently and then have a list of links to https://try.mudblazor.com/ to exemplify all of the features and real-world use cases. Having a single page for documentation on a component like this does not really suffice and will result in a massive page eventually. Take a look at the documentation for DataGrid in MUI for a reference. They have multiple pages for DataGrid docs, each dealing with a particular concept. https://mui.com/components/data-grid/

is this ready for merge? (just need to solve the conflic i just did some structure changes to the solution and renamed Client/Server -> Wasm,WasmHost and added Blazor server as "Server"

When its merged i take a look at it docs wise, but unless i remember wrong i suggested to you way back that we group it under "Table" with several pages.

tjscience commented:
> > @Garderoben Due to the breadth of functionality available in the data grid, I think that docs-wise it would be better to keep the small number of showcase examples that exist currently and then have a list of links to https://try.mudblazor.com/ to exemplify all of the features and real-world use cases. Having a single page for documentation on a component like this does not really suffice and will result in a massive page eventually. Take a look at the documentation for DataGrid in MUI for a reference. They have multiple pages for DataGrid docs, each dealing with a particular concept. https://mui.com/components/data-grid/
> 
> is this ready for merge? (just need to solve the conflic i just did some structure changes to the solution and renamed Client/Server -> Wasm,WasmHost and added Blazor server as "Server"
> 
> When its merged i take a look at it docs wise, but unless i remember wrong i suggested to you way back that we group it under "Table" with several pages.

It is ready for merge. 

I do remember something like that. Should be good.

mckaragoz commented:
Do we have a Phase 3?

tjscience commented:
> Do we have a Phase 3?

Yes, https://github.com/tjscience/MudBlazor/projects/3
